File: backend/app/Models/MonitoringModel.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class MonitoringModel extends Model
{
    public $incrementing = false;
    protected $keyType = 'string';
    protected $table = 'monitoring_models';

    protected $fillable = [
        'id',
        'pengajuan_id',
        'survey_schedule_id',
        'nomor_surat_ba',
        'tanggal_monitoring',
        'nama_perumahan',
        'nama_developer',
        'lokasi_perumahan',
        'maksud_tujuan',
        'temuan_lapangan',
        'kesimpulan',
        'kesepakatan',
        'rencana_tindak_lanjut',
        'status_hasil_evaluasi',
        'pelaksana_nama',
        'pelaksana_jabatan',
        'ditemui_nama',
        'ditemui_jabatan',
        'photo_paths',
        'is_draft',
        'created_by',
    ];

    protected $casts = [
        'tanggal_monitoring' => 'datetime',
        'temuan_lapangan' => 'array',
        'kesimpulan' => 'array',
        'kesepakatan' => 'array',
        'rencana_tindak_lanjut' => 'array',
        'photo_paths' => 'array',
        'is_draft' => 'boolean',
    ];

    public function pengajuan()
    {
        return $this->belongsTo(Pengajuan::class, 'pengajuan_id');
    }

    public function surveySchedule()
    {
        return $this->belongsTo(SurveySchedule::class, 'survey_schedule_id');
    }

    /**
     * Format response JSON yang PERSIS sama dengan kebutuhan FE Flutter
     */
    public function toFlutterArray(): array
    {
        return [
            'id' => $this->id,
            'pengajuanId' => $this->pengajuan_id,
            'surveyScheduleId' => $this->survey_schedule_id,
            'nomorSuratBA' => $this->nomor_surat_ba,
            'tanggalMonitoring' => $this->tanggal_monitoring ? $this->tanggal_monitoring->toIso8601String() : null,
            'namaPerumahan' => $this->nama_perumahan,
            'namaDeveloper' => $this->nama_developer,
            'lokasiPerumahan' => $this->lokasi_perumahan,
            'maksudTujuan' => $this->maksud_tujuan,
            'temuanLapangan' => $this->temuan_lapangan ?? [],
            'kesimpulan' => $this->kesimpulan ?? [],
            'kesepakatan' => $this->kesepakatan ?? [],
            'rencanaTindakLanjut' => $this->rencana_tindak_lanjut ?? [],
            'statusHasilEvaluasi' => $this->status_hasil_evaluasi,
            'pelaksanaNama' => $this->pelaksana_nama,
            'pelaksanaJabatan' => $this->pelaksana_jabatan,
            'ditemuiNama' => $this->ditemui_nama,
            'ditemuiJabatan' => $this->ditemui_jabatan,
            'photoPaths' => $this->photo_paths ?? [],
            'isDraft' => (bool) $this->is_draft,
        ];
    }
}

File: lib/features/monitoring/data/models/monitoring_model.dart
import 'status_hasil_evaluasi.dart';

class MonitoringModel {
  static const String defaultMaksudTujuan =
      'Kegiatan ini dilaksanakan guna memastikan kesesuaian pelaksanaan dengan '
      'perencanaan sebagai bagian dari fungsi pengawasan dan pengendalian '
      'terhadap proses pembangunan perumahan beserta penyediaan PSU-nya.';

  static const List<String> _bulanRomawi = [
    'I', 'II', 'III', 'IV', 'V', 'VI',
    'VII', 'VIII', 'IX', 'X', 'XI', 'XII'
  ];

  /// Generate nomor surat BA: 600.2.5/BA-MON/[bulanRomawi]/[tahun]
  static String generateNomorSuratBA(DateTime tanggal) {
    final bulan = _bulanRomawi[tanggal.month - 1];
    final tahun = tanggal.year;
    return '600.2.5/BA-MON/$bulan/$tahun';
  }

  final String id;
  final String? pengajuanId;
  final DateTime tanggalMonitoring;
  final String namaPerumahan;
  final String namaDeveloper;
  final String lokasiPerumahan;
  final String maksudTujuan;
  final List<String> temuanLapangan;
  final List<String> kesimpulan;
  final List<String> kesepakatan;
  final List<String> rencanaTindakLanjut;
  final StatusHasilEvaluasi statusHasilEvaluasi;
  final String pelaksanaNama;
  final String pelaksanaJabatan;
  final String ditemuiNama;
  final String ditemuiJabatan;
  final List<String> photoPaths;
  final DateTime createdAt;
  final String nomorSuratBA;
  final bool isDraft;

  MonitoringModel({
    required this.id,
    this.pengajuanId,
    required this.tanggalMonitoring,
    required this.namaPerumahan,
    this.namaDeveloper = '',
    this.lokasiPerumahan = '',
    this.maksudTujuan = defaultMaksudTujuan,
    this.temuanLapangan = const [],
    this.kesimpulan = const [],
    this.kesepakatan = const [],
    this.rencanaTindakLanjut = const [],
    this.statusHasilEvaluasi = StatusHasilEvaluasi.sesuaiSiteplan,
    this.pelaksanaNama = '',
    this.pelaksanaJabatan = '',
    this.ditemuiNama = '',
    this.ditemuiJabatan = '',
    this.photoPaths = const [],
    DateTime? createdAt,
    String? nomorSuratBA,
    this.isDraft = false,
  })  : createdAt = createdAt ?? DateTime.now(),
        nomorSuratBA = nomorSuratBA ?? generateNomorSuratBA(tanggalMonitoring);

  MonitoringModel copyWith({
    String? id,
    String? pengajuanId,
    DateTime? tanggalMonitoring,
    String? namaPerumahan,
    String? namaDeveloper,
    String? lokasiPerumahan,
    String? maksudTujuan,
    List<String>? temuanLapangan,
    List<String>? kesimpulan,
    List<String>? kesepakatan,
    List<String>? rencanaTindakLanjut,
    StatusHasilEvaluasi? statusHasilEvaluasi,
    String? pelaksanaNama,
    String? pelaksanaJabatan,
    String? ditemuiNama,
    String? ditemuiJabatan,
    List<String>? photoPaths,
    DateTime? createdAt,
    String? nomorSuratBA,
    bool? isDraft,
  }) {
    return MonitoringModel(
      id: id ?? this.id,
      pengajuanId: pengajuanId ?? this.pengajuanId,
      tanggalMonitoring: tanggalMonitoring ?? this.tanggalMonitoring,
      namaPerumahan: namaPerumahan ?? this.namaPerumahan,
      namaDeveloper: namaDeveloper ?? this.namaDeveloper,
      lokasiPerumahan: lokasiPerumahan ?? this.lokasiPerumahan,
      maksudTujuan: maksudTujuan ?? this.maksudTujuan,
      temuanLapangan: temuanLapangan ?? this.temuanLapangan,
      kesimpulan: kesimpulan ?? this.kesimpulan,
      kesepakatan: kesepakatan ?? this.kesepakatan,
      rencanaTindakLanjut: rencanaTindakLanjut ?? this.rencanaTindakLanjut,
      statusHasilEvaluasi: statusHasilEvaluasi ?? this.statusHasilEvaluasi,
      pelaksanaNama: pelaksanaNama ?? this.pelaksanaNama,
      pelaksanaJabatan: pelaksanaJabatan ?? this.pelaksanaJabatan,
      ditemuiNama: ditemuiNama ?? this.ditemuiNama,
      ditemuiJabatan: ditemuiJabatan ?? this.ditemuiJabatan,
      photoPaths: photoPaths ?? this.photoPaths,
      createdAt: createdAt ?? this.createdAt,
      nomorSuratBA: nomorSuratBA ?? this.nomorSuratBA,
      isDraft: isDraft ?? this.isDraft,
    );
  }
}
